File: src/year2019/day25/day25.rs
//! [Day 25: Cryostasis](https://adventofcode.com/2019/day/25)

use intcode::{Computer, State};
use rustc_hash::{FxHashMap, FxHashSet};

type Room = String;
type Direction = String;
type Map = FxHashMap<Room, FxHashMap<Direction, Room>>;

fn run(computer: &mut Computer, command: &str) -> String {
    computer.input_flush();

    if !command.is_empty() {
        command.bytes().for_each(|byte| computer.push_byte(byte));
        computer.push_byte(b'\n');
    }

    let mut output = String::new();
    for _ in 1..10000 {
        match computer.run() {
            State::Output(ch) => {
                output.push(char::from_u32(u32::try_from(ch).unwrap()).unwrap());
            }
            State::Halted | State::Input => return output,
        }
    }

    String::new()
}

fn parse(output: &str) -> (&str, Vec<&str>, Vec<&str>) {
    let mut dirs = Vec::new();
    let mut items = Vec::new();
    let mut room = "";

    for line in output.lines() {
        if let Some(s) = line.strip_prefix("== ") {
            if let Some(s) = s.strip_suffix(" ==") {
                if room.is_empty() {
                    room = s;
                }
            }
        }

        if let Some(s) = line.strip_prefix("- ") {
            match s {
                "north" | "east" | "south" | "west" => dirs.push(s),
                _ => items.push(s),
            };
        }
    }

    (room, dirs, items)
}

fn reverse(direction: &str) -> &'static str {
    match direction {
        "north" => "south",
        "south" => "north",
        "east" => "west",
        "west" => "east",
        _ => "none",
    }
}

fn explore(computer: &mut Computer, map: &mut Map, output: &str) {
    //

    let (room, dirs, items) = parse(output);

    for dir in &dirs {
        if map.contains_key(room) && map[room].contains_key(*dir) {
            continue;
        }

        // go next room
        let output = run(computer, dir);

        let (newroom, _, _) = parse(&output);
        let known = map.contains_key(newroom);

        map.entry(room.to_string())
            .or_default()
            .insert((*dir).to_string(), newroom.to_string());
        map.entry(newroom.to_string())
            .or_default()
            .insert(reverse(dir).to_string(), room.to_string());

        if output.contains("A loud") {
            continue;
        }

        if !known {
            explore(computer, map, &output);
        }

        run(computer, reverse(dir));
    }

    for item in items {
        let mut temp_computer = computer.clone();

        let command = format!("take {item}");

        let output = run(&mut temp_computer, &command);
        if output.is_empty() {
            // stuck
            continue;
        }

        let output = run(&mut temp_computer, "inv");
        if output.contains("Items in your inventory") {
            run(computer, &command);
        }
    }
}

fn find_path<'a>(map: &'a Map, start: &str, target: &str) -> Vec<&'a str> {
    let mut seen = FxHashSet::default();

    let mut stack = Vec::new();

    stack.push((start, Vec::new()));

    while let Some((current, path)) = stack.pop() {
        if current == target {
            return path;
        }

        for (dir, cell) in &map[current] {
            if seen.insert(cell) {
                let mut new_path = path.clone();
                new_path.push(dir);
                stack.push((cell, new_path));
            }
        }
    }

    Vec::new()
}

fn find_weight(computer: &mut Computer, inventory: &[&str], checkpoint_dir: &str) -> u64 {
    let re = regex::Regex::new(
        r"You should be able to get in by typing (\d+) on the keypad at the main airlock.",
    )
    .unwrap();

    let mut prev_code = 0i32;

    for code in 0..(1 << inventory.len()) {
        let gray_code = code ^ (code >> 1);
        let diff = gray_code - prev_code;

        if diff != 0 {
            let action = if diff > 0 { "drop" } else { "take" };

            let mut diff = diff.abs();
            let mut item = 0;
            while diff & 1 == 0 {
                diff >>= 1;
                item += 1;
            }

            let command = format!("{action} {}", inventory[item]);
            run(computer, &command);
        }

        let output = run(computer, checkpoint_dir);

        if let Some(caps) = re.captures(&output) {
            return caps[1].parse().unwrap();
        }

        prev_code = gray_code;
    }

    0
}

/// # Panics
#[must_use]
pub fn solve(program: &str) -> (u64, aoc::Christmas) {
    let mut computer = Computer::load(program);

    let mut map = FxHashMap::default();

    let output = run(&mut computer, "");
    let (start_room, _, _) = parse(&output);

    // visit all rooms and collect items
    explore(&mut computer, &mut map, &output);

    // find path to the Pressure-Sensitive Floor
    let path = find_path(&map, start_room, "Pressure-Sensitive Floor");

    // the direction to go to Security Checkpoint from Pressure-Sensitive Floor
    let checkpoint_dir = map["Security Checkpoint"]
        .iter()
        .filter(|(_, room)| room == &"Pressure-Sensitive Floor")
        .map(|(dir, _)| dir)
        .next()
        .unwrap();

    // go to the Pressure-Sensitive Floor
    for step in path {
        run(&mut computer, step);
    }

    // the inventory
    let inventory = run(&mut computer, "inv");
    let inventory = inventory
        .lines()
        .filter_map(|line| line.strip_prefix("- "))
        .collect::<Vec<&str>>();

    // get the unlock code
    (
        find_weight(&mut computer, &inventory, checkpoint_dir),
        aoc::CHRISTMAS,
    )
}

pub fn main() {
    let args = aoc::parse_args();
    args.run(solve);
}

File: src/year2019/day25/day25.py
#!/usr/bin/env python3
# [Day 25: Cryostasis](https://adventofcode.com/2019/day/25)

import argparse
import atexit
import re
import sys
import time
import typing as t
from collections import defaultdict
from pathlib import Path

sys.path.append(Path(__file__).parent.parent.as_posix())
from intcode.Intcode import Computer  # noqa


def run(computer, command=None):

    computer.flush_io()

    if command:
        computer.input.extend(map(ord, command))
        computer.input.append(10)

    computer.resume()

    return "".join(map(chr, computer.output))


def parse(out: str) -> t.Tuple[str, t.List, t.List]:

    room = None
    dirs = []
    items = []

    for line in out.splitlines():
        if line.startswith("== ") and line.endswith(" =="):
            if not room:
                room = line[3:-3]
        if line.startswith("- "):
            line = line.removeprefix("- ")
            if line in ("north", "east", "south", "west"):
                dirs.append(line)
            else:
                items.append(line)

    return room, dirs, items


def reverse(dir: str) -> str:
    match dir:
        case "north":
            return "south"
        case "south":
            return "north"
        case "east":
            return "west"
        case "west":
            return "east"

    raise ValueError(dir)


def explore(computer, map, out):

    room, dirs, items = parse(out)

    for dir in dirs:
        if dir in map[room]:
            continue

        # go next room
        out = run(computer, dir)

        newroom, _, _ = parse(out)
        known = newroom in map

        map[room][dir] = newroom
        map[newroom][reverse(dir)] = room

        if "A loud" in out:
            continue

        if not known:
            explore(computer, map, out)

        # go back
        out = run(computer, reverse(dir))

    for item in items:

        cmd = f"take {item}"

        temp_computer = computer.clone()
        temp_computer.max_iterations = 10000  # prevent infinite loop
        run(temp_computer, cmd)

        if temp_computer._state == "read":
            out = run(temp_computer, "inv")
            if "Items in your inventory" in out:
                run(computer, cmd)


def find_path(map, current, target, seen=set()) -> t.List[str]:

    seen.add(current)
    if current == target:
        return []

    for dir, cell in map[current].items():
        if cell not in seen:
            if (path := find_path(map, cell, target, seen)) is not None:
                return [dir] + path


def find_weight(computer, inventory, checkpoint_dir) -> str:

    prev_code = 0
    for cod in range(1 << len(inventory)):

        gray_code = cod ^ (cod >> 1)

        diff = gray_code - prev_code

        if diff != 0:
            cmd = "drop" if diff > 0 else "take"
            diff = abs(diff)
            item = 0
            while diff & 1 == 0:
                diff >>= 1
                item += 1
            run(computer, f"{cmd} {inventory[item]}")

        out = run(computer, checkpoint_dir)
        answer = re.search(r"You should be able to get in by typing (\d+) on the keypad at the main airlock.", out)
        if answer:
            return answer[1]

        prev_code = gray_code


def solve(software):

    computer = Computer()
    computer.load(software)
    computer.start()

    map = defaultdict(dict)

    # start up
    out = run(computer)
    start_room, _, _ = parse(out)

    # visit all rooms and take items
    explore(computer, map, out)

    # find path to the detection room
    path = find_path(map, start_room, "Pressure-Sensitive Floor")

    # the direction to get to Security Checkpoint from Pressure-Sensitive Floor
    checkpoint_dir = [k for k, v in map["Security Checkpoint"].items() if v == "Pressure-Sensitive Floor"][0]

    # go to the Pressure-Sensitive Floor
    for step in path:
        run(computer, step)

    # the inventory
    out = run(computer, "inv")
    inventory = [s.removeprefix("- ") for s in out.splitlines() if s.startswith("- ")]

    # get the unlock code
    unlock = find_weight(computer, inventory, checkpoint_dir)

    print(unlock)


def main():

    parser = argparse.ArgumentParser()
    parser.add_argument("-v", "--verbose", action="store_true")
    parser.add_argument("--elapsed", action="store_true")
    parser.add_argument("input", nargs="?", default="input.txt")
    args = parser.parse_args()

    software = Path(args.input).read_text()

    if args.elapsed:
        start_time_ns = time.time_ns()
        atexit.register(lambda: print(f"elapsed: {(time.time_ns() - start_time_ns) / 1_000_000}ms"))

    solve(software)


if __name__ == "__main__":
    main()
